### ISSUE TITLE
Explainer Index out of range

### ISSUE THREAD

**Issue by LukasFides** [shap/shap#2617]
Hi, I am trying out this great framework with a self trained GPT-2.

I wanted to use a custom trained model and the base model as tokenizer.

No matter if I use this approach or solely the base model ("dbmdz/german-gpt2") I receive the following error:

```
[/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py](https://dbhohk3q2ad-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220711-060047-RC00_460172213#) in embedding(input, weight, padding_idx, max_norm, norm_type, scale_grad_by_freq, sparse)
   2197         # remove once script supports set_grad_enabled
   2198         _no_grad_embedding_renorm_(weight, input, max_norm, norm_type)
-> 2199     return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
   2200 
   2201 

IndexError: index out of range in self
```

Here is my code:

```
import numpy as np
from transformers import AutoTokenizer, AutoModelForCausalLM
import shap
import torch

model_name = "dbmdz/german-gpt2"
model_n = "dbmdz/german-gpt2"

tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True)
model = AutoModelForCausalLM.from_pretrained("/content/gdrive/MyDrive/Models/gpt2-contract_40epoch") 
#or model = AutoModelForCausalLM.from_pretrained(model_name) 

 set model decoder to true
model.config.is_decoder=True
#set text-generation params under task_specific_params
model.config.task_specific_params["text-generation"] = {
    "do_sample": True,
    "max_length": 50,
    "temperature": 0.7,
    "top_k": 50,
    "no_repeat_ngram_size": 2
}

#sample for generation

s = ['I enjoy walking with my cute dog']
explainer = shap.Explainer(model, tokenizer)
shap_values = explainer(s)

shap.plots.text(shap_values)
```

**Comment by remo-help** [comment]
Hi Lukas, 
 can you please post the entire error log so we can see which line triggered this error. Without this it's unclear whether this is a shap issue.

Can you run `model.predict(s)`?

**Comment by LukasFides** [comment]
Hi @remo-help and thank you for your help.
I can not call `model.predict(s)`:

`GPT2LMHeadModel' object has no attribute 'predict'`

But I can generate text with a transformer pipeline as follows:

`gpt = pipeline('text-generation',model=model, tokenizer=tokenizer,config={'max_length':1000})`

Also, here is the complete error log:

```

---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
[<ipython-input-13-2cc8a2203102>](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in <module>()
      1 explainer = shap.Explainer(model, tokenizer)
----> 2 shap_values = explainer(s)
      3 
      4 shap.plots.text(shap_values)

15 frames
[/usr/local/lib/python3.7/dist-packages/shap/explainers/_partition.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in __call__(self, max_evals, fixed_context, main_effects, error_bounds, batch_size, outputs, silent, *args)
    136         return super().__call__(
    137             *args, max_evals=max_evals, fixed_context=fixed_context, main_effects=main_effects, error_bounds=error_bounds, batch_size=batch_size,
--> 138             outputs=outputs, silent=silent
    139         )
    140 

[/usr/local/lib/python3.7/dist-packages/shap/explainers/_explainer.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in __call__(self, max_evals, main_effects, error_bounds, batch_size, outputs, silent, *args, **kwargs)
    266             row_result = self.explain_row(
    267                 *row_args, max_evals=max_evals, main_effects=main_effects, error_bounds=error_bounds,
--> 268                 batch_size=batch_size, outputs=outputs, silent=silent, **kwargs
    269             )
    270             values.append(row_result.get("values", None))

[/usr/local/lib/python3.7/dist-packages/shap/explainers/_partition.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in explain_row(self, max_evals, main_effects, error_bounds, batch_size, outputs, silent, fixed_context, *row_args)
    182         self.dvalues = np.zeros(out_shape)
    183 
--> 184         self.owen(fm, self._curr_base_value, f11, max_evals - 2, outputs, fixed_context, batch_size, silent)
    185 
    186         # if False:

[/usr/local/lib/python3.7/dist-packages/shap/explainers/_partition.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in owen(self, fm, f00, f11, max_evals, output_indexes, fixed_context, batch_size, silent)
    294             # run the batch
    295             if len(batch_args) > 0:
--> 296                 fout = fm(batch_masks)
    297                 if output_indexes is not None:
    298                     fout = fout[:,output_indexes]

[/usr/local/lib/python3.7/dist-packages/shap/utils/_masked_model.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in __call__(self, masks, zero_index, batch_size)
     65 
     66         else:
---> 67             return self._full_masking_call(masks, batch_size=batch_size)
     68 
     69     def _full_masking_call(self, masks, zero_index=None, batch_size=None):

[/usr/local/lib/python3.7/dist-packages/shap/utils/_masked_model.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in _full_masking_call(self, masks, zero_index, batch_size)
    142 
    143             joined_masked_inputs = tuple([np.concatenate(v) for v in all_masked_inputs])
--> 144             outputs = self.model(*joined_masked_inputs)
    145             _assert_output_input_match(joined_masked_inputs, outputs)
    146             all_outputs.append(outputs)

[/usr/local/lib/python3.7/dist-packages/shap/models/_teacher_forcing.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in __call__(self, X, Y)
    111             X_batch = X[start_batch_idx:start_batch_idx+self.batch_size]
    112             Y_batch = Y[start_batch_idx:start_batch_idx+self.batch_size]
--> 113             logits = self.get_teacher_forced_logits(X_batch, Y_batch)
    114             logodds = self.get_logodds(logits)
    115             if output_batch is None:

[/usr/local/lib/python3.7/dist-packages/shap/models/_teacher_forcing.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in get_teacher_forced_logits(self, X, Y)
    360             inputs = self.get_inputs(X, padding_side='left')
    361             # generate outputs and logits
--> 362             logits = self.model_inference(inputs, output_ids)
    363             # extract only logits corresponding to target sentence ids
    364             logits = logits[:, -output_ids.shape[1]-1:-1, :]

[/usr/local/lib/python3.7/dist-packages/shap/models/_teacher_forcing.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in model_inference(self, inputs, output_ids)
    274                     inputs["position_ids"].masked_fill_(inputs["attention_mask"] == 0, 0)
    275                     # model inference
--> 276                     outputs = self.similarity_model(**inputs, return_dict=True)
    277                 logits = outputs.logits.detach().cpu().numpy().astype('float64')
    278         elif self.similarity_model_type == "tf":

[/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in _call_impl(self, *input, **kwargs)
   1128         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1129                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1130             return forward_call(*input, **kwargs)
   1131         # Do not call functions when jit is used
   1132         full_backward_hooks, non_full_backward_hooks = [], []

[/usr/local/lib/python3.7/dist-packages/transformers/models/gpt2/modeling_gpt2.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in forward(self, input_ids, past_key_values, attention_mask, token_type_ids, position_ids, head_mask, inputs_embeds, encoder_hidden_states, encoder_attention_mask, labels, use_cache, output_attentions, output_hidden_states, return_dict)
   1059             output_attentions=output_attentions,
   1060             output_hidden_states=output_hidden_states,
-> 1061             return_dict=return_dict,
   1062         )
   1063         hidden_states = transformer_outputs[0]

[/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in _call_impl(self, *input, **kwargs)
   1128         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1129                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1130             return forward_call(*input, **kwargs)
   1131         # Do not call functions when jit is used
   1132         full_backward_hooks, non_full_backward_hooks = [], []

[/usr/local/lib/python3.7/dist-packages/transformers/models/gpt2/modeling_gpt2.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in forward(self, input_ids, past_key_values, attention_mask, token_type_ids, position_ids, head_mask, inputs_embeds, encoder_hidden_states, encoder_attention_mask, use_cache, output_attentions, output_hidden_states, return_dict)
    832 
    833         if inputs_embeds is None:
--> 834             inputs_embeds = self.wte(input_ids)
    835         position_embeds = self.wpe(position_ids)
    836         hidden_states = inputs_embeds + position_embeds

[/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in _call_impl(self, *input, **kwargs)
   1128         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1129                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1130             return forward_call(*input, **kwargs)
   1131         # Do not call functions when jit is used
   1132         full_backward_hooks, non_full_backward_hooks = [], []

[/usr/local/lib/python3.7/dist-packages/torch/nn/modules/sparse.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in forward(self, input)
    158         return F.embedding(
    159             input, self.weight, self.padding_idx, self.max_norm,
--> 160             self.norm_type, self.scale_grad_by_freq, self.sparse)
    161 
    162     def extra_repr(self) -> str:

[/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py](https://3e29sc6bmb4-496ff2e9c6d22116-0-colab.googleusercontent.com/outputframe.html?vrz=colab-20220713-060046-RC00_460676560#) in embedding(input, weight, padding_idx, max_norm, norm_type, scale_grad_by_freq, sparse)
   2197         # remove once script supports set_grad_enabled
   2198         _no_grad_embedding_renorm_(weight, input, max_norm, norm_type)
-> 2199     return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)
   2200 
   2201 

IndexError: index out of range in self
```

**Comment by remo-help** [comment]
Hi Lukas, I can look at this in detail tomorrow, ~~but I can tell you at a quick glance that the problem is the way you've set this up and not shap itself.~~

You are using a pipeline, but you are passing shap the model. Your model cannot process a string of text, it needs to be tokenized and ~~vectorized~~ first, which is why you have a pipeline.

Equally, shap cannot take a string of text, it needs a 2D array, which is what you should have after vectorization.

You need to tokenize ~~and vectorize~~ your text before you pass it to shap.

**Comment by LukasFides** [comment]
Hi @remo-help ,

Thank you for your help. So in order to use any other GPT-2 with Shap I need to tokenise the string input first? What kind of vectorisation and tokenisation do you recommend?

I tried `encoded = tokenizer.batch_encode_plus(s, return_tensors='pt', add_special_tokens=True)` and still had the same error.

**Comment by remo-help** [comment]
Hi Lukas,
since you're using a transformer, you may not need a vectorizer, since your model does the embedding for you. Equally, you already have a tokenizer and you are calling it in your explainer. I missed that when i just gave this a cursory glance.

Looking at the shap documentation, it looks to me like you are trying the correct approach for a transformer model. What might be happening is that shap does not like to predict to one row of data.

Try either:
`shap_values = explainer(s, fixed_context=1)`
Or:
`s = ['I enjoy walking with my cute dog', 'I enjoy walking my cat']`
and leave the rest of your code as you had it when you posted the issue.

Let me know if this solves your issue.

**Comment by LukasFides** [comment]
Hi @remo-help ,

Thank you for your suggestions. Unfortunately, either adding `fixed_context=1` or changing the text to `s = ['I enjoy walking with my cute dog', 'I enjoy walking my cat']` still leads to `IndexError: index out of range in self`

**Comment by remo-help** [comment]
Hi Lukas, I see you're trying to emulate this [example](https://shap.readthedocs.io/en/latest/example_notebooks/text_examples/text_generation/Open%20Ended%20GPT2%20Text%20Generation%20Explanations.html) with your own model.

Can you run this example without issue?

I looked at your error code and it looks to me like the error comes from your model, specifically when the model generates the embeddings.

Try a model agnostic explainer, to see if this issue persists over other shap explainers:

`explainer = shap.explainers.Permutation(model, tokenizer)`

`shap_values = explainer(s)`

I tried the above codesnippet with the API example and it worked.

And make sure you are running `shap==0.41`

**Comment by LukasFides** [comment]
Hi @remo-help ,

Thank you for your help :)

The code works when I am using "gpt2" as tokenizer. This unfortunately leads to unusable output as my custom models are in German and are trained on the German gpt2 model "dbmdz/german-gpt2". 

If the tokenizer "dbmdz/german-gpt2" is used no matter which model I am using as model ("dbmdz/german-gpt2", or 2 other models trained on "dbmdz/german-gpt2"), I receive an index out of range error.

Also using `explainer = shap.explainers.Permutation(model,tokenizer)` did not solve it unfortunately.

**Comment by remo-help** [comment]
the issue is then your German model, but I'm not sure why

This should not matter with a transformer, but just to be sure, try

`s = ['Ich gehe mit meinem Hund spazieren']`

I would also try if using an English tokenizer solves the issue (leave the model as is)

**Comment by LukasFides** [comment]
Hi @remo-help ,

Thank you for your reply.

Yes I also tried a German input, but the input appears to not affect the outcome. 

Every German model I tried is causing `the index out of range error`.

I also tried (just because I am a bit lost) a Spanish model `DeepESP/gpt2-spanish` and it was able to give an output for the string:
`s = ['Ich gehe mit meinem Hund spazieren']`

The Spanish and English GPT-2s I used both provided an output but of course the output is rather polluted.

Happy for any leads as always.

**Comment by remo-help** [comment]
@ me here next Monday if you haven't solved this by then. I don't have time to look more into it until then. This is a very strange case.

**Comment by LukasFides** [comment]
Hi @remo-help unfortunately I am still facing this and the [BART Abstractive Summarisation](https://github.com/slundberg/shap/issues/2618) Issue with German models. Any help is highly appreciated.

**Comment by remo-help** [comment]
Lukas is the model you use available somewhere so I can try and reproduce your error?

**Comment by LukasFides** [comment]
Hi @remo-help 

Yes, for example [this model](https://huggingface.co/brjezierski/german-gpt2-easy).

Thank you a lot for your help :)

**Comment by remo-help** [comment]
I'm busy the next few days. I'll try and replicate your error on Thursday, hopefully that will give us some more insight as to what's happening here.

**Comment by LukasFides** [comment]
Hi @remo-help did you have had time to watch into the issue. Any help on this or the [BART problem](https://github.com/slundberg/shap/issues/2618) is highly appreciated.

Cheers.
Lukas

**Comment by remo-help** [comment]
Hi @LukasFides I played around with this quite a bit and it's clearly a model specific issue, but I don't understand what the issue is. I can generate shap values just fine using the below code with the normal 'gpt2' model, but also with the following German gpt2 models:
"josh-oo/german-gpt2-easy"
'benjamin/gpt2-wechsel-german'

Why some gpt2 models work with this code and others don't, is beyond me. I tried various workarounds such as different explainers or using the pipeline, but they did not make a difference. 

I used the following environment:
`python == 3.8.13`

`shap == 0.41.0`
`transformers == 4.19.2`
`torch == 1.11`

Here's the code:
```
import numpy as np
from transformers import AutoTokenizer, AutoModelForCausalLM
import shap
import torch


# works: 
# model_name = "josh-oo/german-gpt2-easy"

model_name = 'benjamin/gpt2-wechsel-german'

# does not work:
# model_name = "stefan-it/german-gpt2-larger"
# model_name = "Tanhim/gpt2-model-de"
# model_name = "dbmdz/german-gpt2"


tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True)
model = AutoModelForCausalLM.from_pretrained(model_name) 

# set model decoder to true

model.config.is_decoder=True
# set text-generation params under task_specific_params
model.config.task_specific_params["text-generation"] = {
    "do_sample": True,
    "model_max_length": 50,
    "temperature": 0.7,
    "top_k": 50,
    "no_repeat_ngram_size": 2
}

s = ['Ich gehe mit meinem Hund spazieren.', 'Und mit meinem Freund']

explainer = shap.Explainer(model, tokenizer)
shap_values = explainer(s)

shap.plots.text(shap_values)

```

**Comment by LukasFides** [comment]
Hi @remo-help 

Thank you for your help. When I install the specific package versions in Google Collab and run the two working models I receive an error: 
```

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[<ipython-input-4-9331951002be>](https://localhost:8080/#) in <module>
     36 shap_values = explainer(s)
     37 
---> 38 shap.plots.text(shap_values)

[/usr/local/lib/python3.7/dist-packages/shap/plots/_text.py](https://localhost:8080/#) in text(shap_values, num_starting_labels, grouping_threshold, separator, xmin, xmax, cmax, display)
    212         cmax_computed = None
    213 
--> 214         for i in range(shap_values.shape[-1]):
    215             for j in range(shap_values.shape[0]):
    216                 values, clustering = unpack_shap_explanation_contents(shap_values[j,:,i])

TypeError: 'NoneType' object cannot be interpreted as an integer
```

Here is the link to the [Google Colab](https://colab.research.google.com/drive/12CkejR_yghHWrQNMKybB0gabePHy0RHi?usp=sharing)

**Comment by remo-help** [comment]
I can replicate this error in colab. I'm not sure why it's happening. I've played around with this for about two hours and here's what I found:

- I can only run the above code in a specific conda environment.
- I can clone the environment using the `conda --clone` flag. If I do this, I can run the code.
- If I try to export the environment to a .yml file and install all the identical packages, then I get the same error as you.
- How or why this is happening is beyond my understanding.

In short: I posses a magical conda environment that allows me to run the code I posted. Any attempt to copy this environment yields different results

**Comment by LukasFides** [comment]
Thank you for getting back. Weird. So basically there is no way to get the magical Conda environment I assume :D

**Comment by remo-help** [comment]
There's two ways to replicate a conda environment that I know.
1.  use the `--clone` command, this replicates the environment, but it's only available on the local machine you're working on
2. use the `export` command, this exports a list of packages and their versions as a text file and allows you to rebuild a conda environment from that list. This route allows sharing conda environments between different machines easily.

If I use method 1., then everything works. If I use method 2., the code does not work. So yeah, I can't think of a good way to replicate this result unfortunately. It's quite odd.

**Comment by LukasFides** [comment]
I see, but there is no way to get the specifications that you are using to my local machine in order to get the two gpt models working with shap, am I right? Sorry bit confused here :,)
